### PR TITLE
Fixes #802, #803 and #3775 - genscript issues

### LINKED ISSUE #802: Documentation/implementation mismatch for --genScript flag
The [documentation](http://nimrod-lang.org/nimrodc.html#command-line-switches) says that `--genScript` generates a compile script in the 'nimcache' directory, but it actually generates it in the project root directory.

Which is the intended behavior? Happy to submit a pull request either way.

### LINKED ISSUE #803: --genScript doesn't include nimbase.h file
Generating a compilation script with `--genScript` doesn't work out-of-the-box because the generated compilation commands don't include a path to the nimrod/lib directory.

``` bash
$ cat compile_hello.sh
clang -c  -w  -o nimrod_hello.o nimrod_hello.c
clang -c  -w  -o nimrod-repo_system.o nimrod-repo_system.c
clang   -o hello  nimrod-repo_system.o nimrod_hello.o -ldl
$ sh compile_hello.sh
nimrod_hello.c:5:10: fatal error: 'nimbase.h' file not found
#include "nimbase.h"
         ^
1 error generated.
nimrod-repo_system.c:5:10: fatal error: 'nimbase.h' file not found
#include "nimbase.h"
         ^
```

If I'm correct in assuming that `--genScript` is intended to generate self-contained compileable C source, should it perhaps bundle the required header files into nimcache? Being new to nimrod, I'm not sure if nimbase.h is the only dependency or if there could potentially be others from nimrod/lib.

### LINKED ISSUE #3775: --genScript is broken?
```
$ cat test.nim
echo "Hello"

$ nim c -d:release --genScript test
Hint: system [Processing]
Hint: test [Processing]
CC: test
CC: stdlib_system
Error: execution of an external compiler program 'gcc -c  -w -O3 -fno-strict-aliasing  -o stdlib_system.o stdlib_system.c' failed with exit code: 1024

gcc: error: stdlib_system.c: No such file or directory
gcc: fatal error: no input files
compilation terminated.
```

Ok, let's take a tambourine.

```
$ nim c -c -d:release --genScript test
Hint: system [Processing]
Hint: test [Processing]
Hint: operation successful (9913 lines compiled; 0.137 sec total; 10.102MB; Release Build) [SuccessX]

$ mv compile_test.sh nimcache/

$ cd nimcache/

$ sh compile_test.sh 
test.c:6:21: fatal error: nimbase.h: No such file or directory
 #include "nimbase.h"
                     ^
compilation terminated.
stdlib_system.c:6:21: fatal error: nimbase.h: No such file or directory
 #include "nimbase.h"
                     ^
compilation terminated.
gcc: error: test.o: No such file or directory
```

### PR DESCRIPTION
**Pull request by genotrance**
Fixes #802 - moved compile script into the `nimcache` directory - makes most sense since the idea is to take this `nimcache` directory onto the target machine to build so having the script together with the code makes sense. Also easy since the `.c` file paths don't need any adjustments in the generated script.

Fixes #803 - copies `nimbase.h` into the `nimcache` directory since it is required for build on the target machine.

Fixes #3775 - `--genscript` means `noAbsolutePaths()` so compile commands all failed due to bad paths. But there's no reason to compile anything on this machine, so `--genscript` should imply `--compileOnly`.

### CODE REVIEW

**Comment by Araq** [comment]
Very nice, but it needs a test case to keep this feature from bit-rotting.

**Comment by genotrance** [comment]
Added a new test category for `--genscript` as discussed on IRC. It runs on Windows and Linux.

File: tests/testament/categories.nim [hunk 1]
```diff
@@ -63,6 +64,52 @@ proc compileRodFiles(r: var TResults, cat: Category, options: string) =
...
+  var build = ""
+  var run = ""
+  var cmd = "cd " & flagsDir / "nimcache" & " && "
+  var script = "compile_tgenscript"
+  when defined(windows):
+    cmd = "cmd /c " & cmd & "$#"
+    script &= ".bat"
+
+  when defined(linux):
+    cmd = "bash -c \"" & cmd & "./$#\""
```
Comment (data-man) [inline tests/testament/categories.nim]: And if bash not installed?

Comment (genotrance) [reply tests/testament/categories.nim]: Bash is available on Travis so test passes. Do we need to change this to "sh -c" instead? Do we run full tests on other systems that don't have bash?

File: doc/advopt.txt [hunk 2]
```diff
@@ -35,7 +35,8 @@ Advanced options:
   --noLinking               compile Nim and generated files but do not link
   --noMain                  do not generate a main procedure
   --genScript               generate a compile script (in the 'nimcache'
-                            subdirectory named 'compile_$$project$$scriptext')
+                            subdirectory named 'compile_$project$scriptext'),
```
[the remark this answers is not in the document]
Comment (genotrance) [reply doc/advopt.txt]: Thanks!

File: tests/testament/categories.nim (hunk 1 above)
Comment (genotrance) [reply tests/testament/categories.nim]: I've changed it to sh, makes sense.

File: tests/testament/categories.nim [hunk 3]
```diff
@@ -63,6 +64,52 @@ proc compileRodFiles(r: var TResults, cat: Category, options: string) =
...
+  var script = "compile_tgenscript"
+  when defined(windows):
+    cmd = "cmd /c " & cmd & "$#"
+    script &= ".bat"
+
+  when defined(linux):
+    cmd = "sh -c \"" & cmd & "./$#\""
+    script &= ".sh"
+
+    setFilePermissions(flagsDir / "nimcache" / script,
```
Comment (Araq) [inline tests/testament/categories.nim]: This is not required.

File: tests/testament/categories.nim [hunk 4]
```diff
@@ -63,6 +64,52 @@ proc compileRodFiles(r: var TResults, cat: Category, options: string) =
...
+    if clearCacheFirst: delNimCache(filename, options)
+    testSpec r, makeTest(flagsDir / filename, options, cat)
+
+  # --genscript
+  test "tgenscript", true
+  test "tgenscript"
+
+  var build = ""
+  var run = ""
+  var cmd = "cd " & flagsDir / "nimcache" & " && "
```
Comment (Araq) [inline tests/testament/categories.nim]: Don't use ``&&`` and string munging. Call ``execShellCmd`` once or twice instead.

File: tests/testament/categories.nim [hunk 5]
```diff
@@ -63,6 +64,52 @@ proc compileRodFiles(r: var TResults, cat: Category, options: string) =
...
+    setFilePermissions(flagsDir / "nimcache" / script,
+      {fpUserExec, fpUserWrite, fpUserRead, fpGroupExec, fpGroupRead,
+       fpOthersExec, fpOthersRead})
+
+  build = cmd % script
+  run = cmd % "tgenscript"
+
+  # Build
+  var (outp, errC) = execCmdEx(build)
+  r.total += 1
```
Comment (Araq) [inline tests/testament/categories.nim]: There should be procs for this, don't mess manually with r.total and r.passed here.

**Comment by genotrance** [comment]
Per feeback from @araq, I've improved the testing of `--genscript`. This includes adding a new `testExec()` proc in `testament` to run executables and shell scripts as normal test cases. It is also now clear in the logs which test was run and with what parameters.

```
PASS: tgenscript C --genscript                                     (9.35208607 secs)
PASS: tgenscript C sh -c "cd nimcache/tests/flags/tgenscript_7a9eea86db4a06cfbc3b096715a16af8 && sh compile_tgenscript.sh" (13.59303904 secs)
PASS: tgenscript C nimcache/tests/flags/tgenscript_7a9eea86db4a06cfbc3b096715a16af8/tgenscript (0.27823400 secs)
```

File: tests/testament/tester.nim [hunk 6]
```diff
@@ -404,6 +404,21 @@ proc testC(r: var TResults, test: TTest) =
     if exitCode != 0: given.err = reExitCodesDiffer
   if given.err == reSuccess: inc(r.passed)
 
+proc testExec(r: var TResults, test: TTest) =
```
Comment (Araq) [inline tests/testament/tester.nim]: It is still a bit dubious why this is required, but it's much better than before. Will merge once the tests are green.

**Pull request merged**